src/Workspaces/Core/Desktop/Workspace/Host/TemporaryStorage/TemporaryStorageServiceFactory.ReferenceCountedDisposable.cs

```diff
@@ -0,0 +1,197 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+#if DEBUG
+using System.Diagnostics;
+#endif
+
+namespace Microsoft.CodeAnalysis.Host
+{
+    internal partial class TemporaryStorageServiceFactory
+    {
+        private sealed class ReferenceCountedDisposable<T> : IDisposable
```

Comment:
If we are implementing a ref-counting scheme that depends on Dispose being called, then i would expect to have a finalizer (potentially debug only) telling us when someone has addref'ed, gotten a REferenceCountededDisposable, but then doesn't end up disposing it.  It seems very easy for someone to make one of these types, potentially forget to dispose it, and then leak a refcount, leading to never letting go of the underlying file.